WDL/runtime/workflow.py

```diff
@@ -604,41 +611,55 @@ def run_local_workflow(
     write_values_json(posix_inputs, os.path.join(run_dir, "inputs.json"), namespace=workflow.name)
 
     state = StateMachine(run_id, run_dir, workflow, posix_inputs)
-
-    try:
-        while state.outputs is None:
-            if _test_pickle:
-                state = pickle.loads(pickle.dumps(state))
-
-            next_call = state.step()
-            if next_call:
-                if isinstance(next_call.callee, Tree.Task):
-                    run_callee = run_local_task
-                elif isinstance(next_call.callee, Tree.Workflow):
-                    run_callee = run_local_workflow
-                else:
-                    assert False
-                _, outputs = run_callee(
-                    next_call.callee,  # pyre-fixme
-                    next_call.inputs,
-                    run_id=next_call.id,
-                    run_dir=os.path.join(run_dir, next_call.id),
-                )
-                state.call_finished(next_call.id, outputs)
-    except Exception as exn:
-        logger.debug(traceback.format_exc())
-        if isinstance(exn, TaskFailure):
-            logger.error("%s failed", getattr(exn, "run_id"))
-        else:
-            msg = ""
-            if hasattr(exn, "job_id"):
-                msg += getattr(exn, "job_id") + " "
-            msg += exn.__class__.__name__
-            if str(exn):
-                msg += ", " + str(exn)
-            logger.error(msg)
-            logger.info("run directory: %s", run_dir)
-        raise
+    with TerminationSignalFlag(logger) as terminating:
+        try:
+            with thread_pool as executor:
+                while state.outputs is None:
+                    if _test_pickle:
+                        state = pickle.loads(pickle.dumps(state))
+
+                    next_call = state.step()
+                    if next_call:
+                        if isinstance(next_call.callee, Tree.Task):
+                            run_callee = run_local_task
+                        elif isinstance(next_call.callee, Tree.Workflow):
+                            run_callee = run_local_workflow
+                        else:
+                            assert False
+                        future = executor.submit(run_callee,
+                                                 next_call.callee,  # pyre-fixme
+                                                 next_call.inputs,
+                                                 run_id=next_call.id,
+                                                 run_dir=os.path.join(run_dir, next_call.id),
+                                                 thread_pool=thread_pool,
+                                                 future_task_map=future_task_map
+                                                 )
+                        future_task_map[future] = next_call.id
+
+                    done_iter = futures.as_completed(future_task_map)
+                    for future in done_iter:
```

Comment:
@MDunitz sorry for taking forever!

1. Looping straight through `done_iter` here means we'll wait for all outstanding calls to complete before submitting any new ones to the thread pool. This would lead to sub-optimal parallelism, because the completion of any one call can potentially unblock starting new ones. We'd like to go back and `state.step()` right away after each call completes, meaning we only want to wait for the *first* element of `done_iter` in each loop iteration. 
1. A knock-on effect is that we'll need some scheme to remove the just-completed future from `future_task_map`. [futures.as_completed](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.as_completed) "Any futures that completed before as_completed() is called will be yielded first."
1. I don't think `future_task_map` should be shared with sub-workflow calls; each workflow can keep track of its outstanding calls separately and the shared thread pool will serve to ensure an appropriate number of tasks are running at any one time. Again though we can defer getting sub-workflow parallelization exactly right for now.

Reply:
Neglected to mention:

There would be an additional, inner while loop scheduling calls until `state.step()` returns `None`, before waiting for a future to complete. That condition indicates that there's nothing more to do at the current moment (but there may be more to do in the future as outstanding calls complete), in contrast to the outer while loop running until `state.outputs is not None` indicating the entire workflow is done.

The principle to keep in mind is that once we try to pop something from `futures.as_completed()`, we're blocked and can do no useful work *for an arbitrarily long period of time* until it answers; therefore, we don't want to do that until we're sure there's nothing else we can do first. From the [state.step contract](https://github.com/chanzuckerberg/miniwdl/blob/e9b5d8cfd7c17f2e1d099930fdfc6c282cc2eca2/WDL/runtime/workflow.py#L209-L220):

> It is NOT necessary to await the
        call's completion before another ``step()`` for the next call; this allows the driver
        to orchestrate multiple calls at once. Indeed, the driver should launch as many calls as
        it can support concurrently, by calling ``step()`` in a loop until getting back ``None``

I think the mechanism for filtering `future_task_map` after future `x` completes will be to keep only those keys `y if y is not x`.